File: typescript/triangle/triangle.ts
export class Triangle {
  sides: number[]
  sidesSet: Set<number>

  constructor(...sides: number[]) {
    this.sides = sides.sort()
    this.sidesSet = new Set(sides)
  }

  get isEquilateral(): boolean {
    return this.valid && this.sidesSet.size === 1
  }

  get isIsosceles(): boolean {
    return this.valid && this.sidesSet.size <= 2
  }

  get isScalene(): boolean {
    return this.valid && this.sidesSet.size === 3
  }

  private get valid(): boolean {
    return this.nonZeroSides && 
           this.satisifiesTriangleInequality
  }

  private get nonZeroSides(): boolean {
    return this.sides.every(s => s > 0)
  }

  private get satisifiesTriangleInequality(): boolean {
    const [a, b, c] = this.sides
    return a + b >= c
  }
}

File: ruby/triangle/triangle.rb
class Triangle
  attr_reader :sides

  def initialize(sides)
    @sides = sides
  end

  def equilateral?
    valid? && @sides.uniq.size == 1
  end

  def isosceles?
    valid? && @sides.uniq.size <= 2
  end

  def scalene?
    valid? && @sides.uniq.size == 3
  end

  def valid?
    a, b, c = sides.sort
    sides.all?(&:positive?) && a + b >= c
  end
end
